File: code/cpp/shortest-word-distance.cc
// 12 ms, 09/04/2017
int shortestDistance(vector<string>& words, string word1, string word2)
{
    int n = words.size(), min_dist = n, i1 = n, i2 = n;
    for (int i = 0; i < n; ++i) {
        if (words[i] == word1 || words[i] == word2) {
            if (words[i] == word1)
                i1 = i;
            else
                i2 = i;
            min_dist = min(min_dist, abs(i1 - i2));
        }
    }
    return min_dist;
}

File: code/python/shortest-word-distance.py
# Date  : 03/18/2018
# Time  : O(n)
# Space : O(n) this implementation; could be make O(1) easily

class Solution(object):
    def shortestDistance(self, words, word1, word2):
        """
        :type words: List[str]
        :type word1: str
        :type word2: str
        :rtype: int
        """
        matches1, matches2 = [], []
        for i, word in enumerate(words):
            if word == word1:
                matches1.append(i)
            if word == word2:
                matches2.append(i)           
        
        min_dist, i, j = float('inf'), 0, 0
        while i < len(matches1) and j < len(matches2):
            min_dist = min(min_dist, abs(matches1[i] - matches2[j]))
            if matches1[i] < matches2[j]:
                i += 1
            else: # matches1[i] > matches2[j]
                j += 1
        
        return min_dist
